File: catalog/appconnections/go/argocdconnector/argocdconnector.go
package argocdconnector

import (
	"crypto/tls"
	"encoding/json"
	"errors"
	"fmt"
	"io/ioutil"
	"net/http"
	"net/url"
	"strings"
)

type BasicAuthentication struct {
	UserName string `json:"userName" yaml:"UserName"`
	Password string `json:"password" yaml:"Password"`
}

type UserDefinedCredentials struct {
	BasicAuthentication BasicAuthentication `json:"basicAuthentication" yaml:"BasicAuthentication"`
}

type LinkedApplications struct {
}

type ArgoCDConnector struct {
	AppURL                 string                  `json:"appURL" yaml:"appURL"`
	AppPort                int                     `json:"appPort" yaml:"appPort"`
	Ipv4Address            string                  `json:"ipv4Address" yaml:"ipv4Address"`
	Ipv6Address            string                  `json:"ipv6Address" yaml:"ipv6Address"`
	UserDefinedCredentials *UserDefinedCredentials `json:"userDefinedCredentials" yaml:"userDefinedCredentials"`
	LinkedApplications     *LinkedApplications     `json:"linkedApplications" yaml:"linkedApplications"`
}

func (thisObj *ArgoCDConnector) Validate() (bool, error) {
	basicAuth := thisObj.UserDefinedCredentials.BasicAuthentication
	token, err := thisObj.getAPIToken(basicAuth.UserName, basicAuth.Password)
	if err != nil {
		return false, err
	}
	isEnabled, err := thisObj.verifyAccount(token)
	if err != nil {
		return false, err
	}
	return isEnabled, nil
}

func (basicauthentication *BasicAuthentication) ValidateAttributes() string {
	var emptyAttributes []string
	errorResultStr := ""
	if basicauthentication.UserName == "" {
		emptyAttributes = append(emptyAttributes, "UserName")
	}
	if basicauthentication.Password == "" {
		emptyAttributes = append(emptyAttributes, "Password")
	}
	if len(emptyAttributes) > 0 {
		errorResultStr = fmt.Sprintf("Invalid Credentials: %s", strings.Join(emptyAttributes, ", ")+" is empty")
	}
	return errorResultStr
}

// getAPIToken gets the API token using the provided username and password
func (thisObj *ArgoCDConnector) getAPIToken(username, password string) (string, error) {
	url := fmt.Sprintf("%s/api/v1/session", thisObj.AppURL)
	method := "POST"
	payload := strings.NewReader(fmt.Sprintf(`{
		"username": "%s",
		"password": "%s"
	}`, username, password))

	// Create a custom HTTP client with TLS config to skip SSL verification
	client := &http.Client{
		Transport: &http.Transport{
			TLSClientConfig: &tls.Config{InsecureSkipVerify: true}, // Disable SSL certificate verification
		},
	}

	req, err := http.NewRequest(method, url, payload)
	if err != nil {
		return "", err
	}
	req.Header.Add("Content-Type", "application/json")

	resp, err := client.Do(req)
	if err != nil {
		return "", handleRequestError(err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return "", handleHTTPError(resp.StatusCode)
	}

	body, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return "", err
	}

	var result map[string]string
	err = json.Unmarshal(body, &result)
	if err != nil {
		return "", err
	}

	token, exists := result["token"]
	if !exists {
		return "", errors.New("token not found in the response")
	}

	return token, nil
}

// verifyAccount checks if the user account is enabled
func (thisObj *ArgoCDConnector) verifyAccount(token string) (bool, error) {
	basicAuth := thisObj.UserDefinedCredentials.BasicAuthentication
	url := fmt.Sprintf("%s/api/v1/account/%s", thisObj.AppURL, basicAuth.UserName)
	method := "GET"

	// Create a custom HTTP client with TLS config to skip SSL verification
	client := &http.Client{
		Transport: &http.Transport{
			TLSClientConfig: &tls.Config{InsecureSkipVerify: true}, // Disable SSL certificate verification
		},
	}

	req, err := http.NewRequest(method, url, nil)
	if err != nil {
		return false, err
	}
	req.Header.Set("Authorization", "Bearer "+token)

	resp, err := client.Do(req)
	if err != nil {
		return false, handleRequestError(err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return false, handleHTTPError(resp.StatusCode)
	}

	body, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return false, err
	}

	var accountData map[string]interface{}
	err = json.Unmarshal(body, &accountData)
	if err != nil {
		return false, err
	}

	enabled, exists := accountData["enabled"].(bool)
	if !exists || !enabled {
		return false, nil
	}

	return true, nil
}

// handleRequestError maps request errors to specific error messages
func handleRequestError(err error) error {
	urlErr, ok := err.(*url.Error)
	if ok && urlErr.Timeout() {
		return errors.New("Timeout error: Invalid URL.")
	}
	return errors.New("Connection error: Invalid URL.")
}

// handleHTTPError maps HTTP status codes to error messages
func handleHTTPError(statusCode int) error {
	switch statusCode {
	case http.StatusUnauthorized:
		return errors.New("Error 401: unauthorized. Invalid username and/or password.")
	case http.StatusForbidden:
		return errors.New("Error 403: forbidden. Invalid username and/or password.")
	case http.StatusNotFound:
		return errors.New("Error 404: Invalid username and/or password.")
	default:
		return fmt.Errorf("http error: status code %d", statusCode)
	}
}

File: catalog/applicationtypes/python/applicationtypes/argocdconnector/argocdconnector.py
import http
import json
from typing import List, Any, Dict
from datetime import datetime, timezone
import requests

class BasicAuthentication:
    user_name: str
    password: str

    def __init__(self, user_name: str, password: str) -> None:
        self.user_name = user_name
        self.password = password

    @staticmethod
    def from_dict(obj) -> 'BasicAuthentication':
        user_name, password = "", ""
        if isinstance(obj, dict):
            user_name = obj.get("UserName", "")
            password = obj.get("Password", "")

        return BasicAuthentication(user_name, password)

    def to_dict(self) -> dict:
        result: dict = {}
        result["UserName"] = self.user_name
        result["Password"] = self.password
        return result

    def validate_attributes(self) -> str:
        emptyAttrs = []
        if not self.user_name:
            emptyAttrs.append("UserName")

        if not self.password:
            emptyAttrs.append("Password")

        return "Invalid Credentials: " + ", ".join(
            emptyAttrs) + " is empty" if emptyAttrs else ""


class UserDefinedCredentials:
    basic_authentication: BasicAuthentication

    def __init__(self, basic_authentication: BasicAuthentication) -> None:
        self.basic_authentication = basic_authentication

    @staticmethod
    def from_dict(obj) -> 'UserDefinedCredentials':
        basic_authentication = None
        if isinstance(obj, dict):
            basic_authentication = BasicAuthentication.from_dict(
                obj.get("BasicAuthentication", None))
        return UserDefinedCredentials(basic_authentication)

    def to_dict(self) -> dict:
        result: dict = {}
        result["BasicAuthentication"] = self.basic_authentication.to_dict()
        return result


class ArgoCDConnector:
    app_url: str
    app_port: int
    user_defined_credentials: UserDefinedCredentials

    def __init__(
            self,
            app_url: str = None,
            app_port: int = None,
            user_defined_credentials: UserDefinedCredentials = None) -> None:
        self.app_url = app_url
        self.app_port = app_port
        self.user_defined_credentials = user_defined_credentials

    @staticmethod
    def from_dict(obj) -> 'ArgoCDConnector':
        app_url, app_port, user_defined_credentials = "", "", None

        if isinstance(obj, dict):
            app_url = obj.get("AppURL", "")
            if not app_url:
                app_url = obj.get("appURL", "")
            if not app_url:
                app_url = obj.get("appurl", "")
            app_port = obj.get("AppPort", 0)
            if not app_port:
                app_port = obj.get("appPort", 0)
            user_defined_credentials_dict = obj.get("UserDefinedCredentials",
                                                    None)
            if user_defined_credentials_dict is None:
                user_defined_credentials_dict = obj.get(
                    "userDefinedCredentials", None)
            if bool(user_defined_credentials_dict):
                user_defined_credentials = UserDefinedCredentials.from_dict(
                    user_defined_credentials_dict)

        return ArgoCDConnector(app_url, app_port, user_defined_credentials)

    def to_dict(self) -> dict:
        result: dict = {}
        result["AppURL"] = self.app_url
        result["AppPort"] = self.app_port
        result[
            "UserDefinedCredentials"] = self.user_defined_credentials.to_dict(
            )

        return result

    def validate(self) -> bool and dict:
        basic_auth = self.user_defined_credentials.basic_authentication
        token, error = self.get_api_token(basic_auth.user_name, basic_auth.password)
        if error:
            return False, error
        is_enabled, error = self.verify_account()
        if error:
            return False, error
        return True, None
    
    def get_api_token(self, username, password):
        argo_cd_url, error = self.get_app_url()
        if error:
            return None, error
        url = f"{argo_cd_url}/api/v1/session"
        payload = json.dumps({
            "username": username,
            "password": password
            })
        
        headers = {
            'Content-Type': 'application/json',
        }
        
        try:
            response = requests.request("POST", url, headers=headers, data=payload, verify=False) # verify is used for local development to bye-pass ssl verification
            response.raise_for_status()
            token = response.json().get("token")

            if token:
                return token, None
            else:
                raise ValueError("Token not found in the response.")

        except requests.exceptions.Timeout:
            return None, "TimeoutError: The request timed out. Please try again."
        except requests.exceptions.ConnectionError:
            return None, "ConnectionError: Failed to connect to the server. Invalid URL."
        except requests.exceptions.HTTPError as http_err:
            status_code = response.status_code
            if status_code == http.HTTPStatus.UNAUTHORIZED:
                return None, "Error 401: Unauthorized. Invalid username and/or password."
            elif status_code == http.HTTPStatus.FORBIDDEN:
                return None, "Error 403: Forbidden. Invalid username and/or password.."
            else:
                return None, f"HTTPError: An HTTP error occurred: {http_err}"
        except requests.exceptions.RequestException as req_err:
            return None, f"RequestException: An error occurred during the request: {req_err}"
        except ValueError as ve:
            return None, f"ValueError: {ve}"

    # https://cd.apps.argoproj.io/swagger-ui#operation/AccountService_GetAccount
    def verify_account(self):
        argo_cd_url, error = self.get_app_url()
        if error:
            return None, error
        basic_auth = self.user_defined_credentials.get("BasicAuthentication")
        username = basic_auth.get("UserName","")
        password = basic_auth.get("Password","")

        url = f"{argo_cd_url}/api/v1/account/{username}"
        token, error = self.get_api_token(username, password)
        if error:
            return False, error
        payload = {}
        headers = {
            "Authorization": f"Bearer {token}"
        }

        try:
            response = requests.request("GET", url, headers=headers, data=payload, verify=False)
            response.raise_for_status()

            account_data = response.json()
            enabled = account_data.get("enabled", False)
            if enabled:
                return True, None
            else:
                return False, None
        
        except requests.exceptions.Timeout:
            return None, "TimeoutError: The request timed out. Please try again."
        except requests.exceptions.ConnectionError:
            return None, "ConnectionError: Failed to connect to the server. Invalid URL."
        except requests.exceptions.HTTPError as http_err:
            status_code = response.status_code
            if status_code == http.HTTPStatus.UNAUTHORIZED:
                return None, "Error 401: Unauthorized. Invalid username and/or password."
            elif status_code == http.HTTPStatus.FORBIDDEN:
                return None, "Error 403: Forbidden. Invalid username and/or password."
            elif status_code == http.HTTPStatus.NOT_FOUND:
                return None, f"Error 404: User '{basic_auth.user_name}' not found."
            else:
                return None, f"HTTPError: An HTTP error occurred: {http_err}"
        except requests.exceptions.RequestException as req_err:
            return None, f"RequestException: An error occurred during the request: {req_err}"
        

    # https://cd.apps.argoproj.io/swagger-ui#operation/ApplicationService_List
    def get_applications(self):
        argo_cd_url, error = self.get_app_url()
        if error:
            return False, error
        basic_auth = self.user_defined_credentials.get("BasicAuthentication")
        username = basic_auth.get("UserName","")
        password = basic_auth.get("Password","")

        url = f"{argo_cd_url}/api/v1/applications"
        token, error = self.get_api_token(username, password)
        if error:
            return False, error
        headers = {
            "Authorization": f"Bearer {token}"
        }
        try:

            response = requests.get(url, headers=headers, verify=False)
            response.raise_for_status()
            
            applications_data = response.json()
            return applications_data, None
        except requests.exceptions.Timeout:
            return None, "TimeoutError: The request timed out. Please try again."
        except requests.exceptions.ConnectionError:
            return None, "ConnectionError: Failed to connect to the server. Invalid URL."
        except requests.exceptions.HTTPError as http_err:
            status_code = response.status_code
            if status_code == http.HTTPStatus.UNAUTHORIZED:
                return None, "Error 401: Unauthorized. Invalid username and/or password."
            elif status_code == http.HTTPStatus.FORBIDDEN:
                return None, "Error 403: Forbidden. Invalid username and/or password."
            elif status_code == http.HTTPStatus.NOT_FOUND:
                return None, "Error 404: Not Found. The endpoint may be incorrect."
            else:
                return None, f"HTTPError: An HTTP error occurred: {http_err}"
        except requests.exceptions.RequestException as req_err:
            return None, f"RequestException: An error occurred during the request: {req_err}"
        
    def get_app_url(self):
        url = self.app_url
        if not url:
            return None, "AppURL is empty"

        return url.rstrip().rstrip("/"), None
    
    def get_current_datetime(self):       
        current_time = datetime.now(timezone.utc)
        formatted_time = current_time.strftime('%Y-%m-%dT%H:%M:%S.%fZ')
        return formatted_time